File: src/main/java/me/lokka30/levelledmobs/wrappers/SchedulerResult.java
package me.lokka30.levelledmobs.wrappers;

import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
import org.bukkit.scheduler.BukkitTask;
import org.jetbrains.annotations.Nullable;

public class SchedulerResult {
    public SchedulerResult(final @Nullable ScheduledTask foliaTask){
        this.foliaTask = foliaTask;
        this.regularTask = null;
        this.isRunningFolia = true;
    }

    public SchedulerResult(final @Nullable BukkitTask regularTask){
        this.regularTask = regularTask;
        this.foliaTask = null;
        this.isRunningFolia = false;
    }

    public final @Nullable ScheduledTask foliaTask;
    public final @Nullable BukkitTask regularTask;
    public final boolean isRunningFolia;

    public void cancelTask(){
        if (isRunningFolia){
            if (foliaTask != null) foliaTask.cancel();
        }
        else{
            if (regularTask != null) regularTask.cancel();
        }
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean isCancelled(){
        if (isRunningFolia){
            if (foliaTask != null)
                return foliaTask.isCancelled();
            else
                return false;
        }
        else{
            if (regularTask != null)
                return regularTask.isCancelled();
            else
                return false;
        }
    }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/logic/SchedulerResult.kt
package io.github.arcaneplugins.levelledmobs.bukkit.logic

import io.papermc.paper.threadedregions.scheduler.ScheduledTask
import org.bukkit.scheduler.BukkitTask

class SchedulerResult(
    val foliaTask: ScheduledTask? = null,
    val regularTask: BukkitTask? = null
) {
    val isRunningFolia = foliaTask != null

    fun cancelTask(){
        if (isRunningFolia) {
            foliaTask?.cancel()
        } else {
            regularTask?.cancel()
        }
    }

    val isCancelled: Boolean
        get() {
            return if (isRunningFolia) {
                foliaTask?.isCancelled ?: false
            } else {
                regularTask?.isCancelled ?: false
            }
        }
}